File: leetcode/tree/easy/java_solutions/ConstructStringFromBinaryTree_606.java
package leetcode.tree.easy.java_solutions;

import static leetcode.tree.Helper.*;

public class ConstructStringFromBinaryTree_606 {

    private static class Solution {
        public String tree2str(TreeNode root) {
            if (root == null) return "";
            String result = root.val + "";
            String left = tree2str(root.left);
            String right = tree2str(root.right);
            if (left.equals("") && right.equals("")) {
                return result;
            } else if (left.equals("")) {
                return result + "()" + "(" + right + ")";
            } else if (right.equals("")) {
                return result + "(" + left + ")";
            }
            return result + "(" + left + ")" + "(" + right + ")";
        }
    }

    private static class SolutionBuilder {
        public String tree2str(TreeNode t) {
            if (t == null) {
                return "";
            }
            StringBuilder sb = new StringBuilder();
            dfs(t, sb);
            return sb.toString();
        }

        private void dfs(TreeNode root, StringBuilder sb) {
            if (root == null) {
                return;
            }
            sb.append(root.val);
            if (root.left != null) {
                sb.append("(");
                dfs(root.left, sb);
                sb.append(")");
            }
            if (root.right != null) {
                if (root.left == null) {
                    sb.append("()");
                }
                sb.append("(");
                dfs(root.right, sb);
                sb.append(")");
            }
        }
    }
}

File: leetcode/tree/easy/golang_solutions/construct_string_from_binary_tree_606.go
package golang_solutions

import (
	"bytes"
	"github.com/vielendanke/algorithms_in_place/leetcode/tree"
	"strconv"
)

var leftBracket = []byte("(")
var rightBracket = []byte(")")

// with omitting

func tree2strOmit(root *tree.TreeNode) string {
	buff := bytes.NewBufferString("")

	dfsOmit(buff, root)

	return buff.String()
}

func dfsOmit(buff *bytes.Buffer, node *tree.TreeNode) {
	if node == nil {
		return
	}
	buff.Write([]byte(strconv.Itoa(node.Val)))

	if node.Left != nil {
		buff.Write(leftBracket)
		dfsOmit(buff, node.Left)
		buff.Write(rightBracket)
	}
	if node.Right != nil {
		if node.Left == nil {
			buff.Write(leftBracket)
			buff.Write(rightBracket)
		}
		buff.Write(leftBracket)
		dfsOmit(buff, node.Right)
		buff.Write(rightBracket)
	}
}

// without omitting

func tree2str(root *tree.TreeNode) string {
	buff := bytes.NewBufferString("")

	dfs(buff, root)

	return buff.String()
}

func dfs(buff *bytes.Buffer, node *tree.TreeNode) {
	if node.Left == nil && node.Right == nil {
		buff.Write([]byte(strconv.Itoa(node.Val)))
		return
	}
	buff.Write([]byte(strconv.Itoa(node.Val)))
	if node.Left != nil {
		buff.Write([]byte(leftBracket))
		dfs(buff, node.Left)
		buff.Write([]byte(rightBracket))
	}
	if node.Right != nil {
		buff.Write([]byte(leftBracket))
		dfs(buff, node.Right)
		buff.Write([]byte(rightBracket))
	}
}
